src/main/java/org/broadinstitute/hellbender/tools/copynumber/utils/segmentation/KernelSegmenter.java

```diff
@@ -0,0 +1,526 @@
+    //and returns the requested number
+    private static List<Integer> selectChangepoints(final List<Integer> changepointCandidates,
+                                                    final int maxNumChangepoints,
+                                                    final double numChangepointsPenaltyLinearFactor,
+                                                    final double numChangepointsPenaltyLogLinearFactor,
+                                                    final RealMatrix reducedObservationMatrix,
+                                                    final double[] kernelApproximationDiagonal) {
+        final List<Integer> changepoints = new ArrayList<>(changepointCandidates.size());
+
+        //calculate penalties as a function of the number of changepoints
+        final int numData = reducedObservationMatrix.getRowDimension();
+        final List<Double> changepointPenalties = IntStream.range(0, maxNumChangepoints + 1).boxed()
+                .map(numChangepoints -> numChangepointsPenaltyLinearFactor * numChangepoints
+                        + numChangepointsPenaltyLogLinearFactor * numChangepoints * Math.log(numData / (numChangepoints + EPSILON)))
+                .collect(Collectors.toList());
+
+        //construct initial list of all segments and initialize costs
+        final List<Integer> candidateStarts = changepointCandidates.stream().sorted().distinct()
+                .map(i -> Math.min(i + 1, numData - 1)).collect(Collectors.toList());
+        candidateStarts.add(0, 0);
+        final List<Integer> candidateEnds = changepointCandidates.stream().sorted().distinct().collect(Collectors.toList());
+        candidateEnds.add(numData - 1);
+        final int numSegments = candidateStarts.size();
+        final List<Segment> segments = IntStream.range(0, numSegments).boxed()
+                .map(i -> new Segment(candidateStarts.get(i), candidateEnds.get(i), reducedObservationMatrix, kernelApproximationDiagonal))
+                .collect(Collectors.toList());
+        final List<Double> totalSegmentationCosts = new ArrayList<>(Collections.singletonList(segments.stream().mapToDouble(s -> s.cost).sum()));
+        final List<Double> costsForSegmentPairs = IntStream.range(0, numSegments - 1).boxed()
+                .map(i -> segments.get(i).cost + segments.get(i + 1).cost)
+                .collect(Collectors.toList());  //sum of the costs for the segments in each adjacent pair
+        final List<Double> costsForMergedSegmentPairs = IntStream.range(0, numSegments - 1).boxed()
+                .map(i -> new Segment(candidateStarts.get(i), candidateEnds.get(i + 1), reducedObservationMatrix, kernelApproximationDiagonal).cost)
+                .collect(Collectors.toList());  //cost of each adjacent pair when considered as a single segment
+        final List<Double> costsForMergingSegmentPairs = IntStream.range(0, numSegments - 1).boxed()
```

Comment:
If you use `double[]` above this reduces to `MathArrays.ebeSubtract`.

Reply:
Good point.  However, I'm removing elements from the lists, so I'd rather use an `ArrayList` (since I like to use arrays to represent fixed-length quantities).